### PR TITLE
Add the ability to break on call-site locations, improve inline stepping

### PR DESCRIPTION
**Pull request by jimingham**
Previously lldb didn't support setting breakpoints on call site locations.  This patch adds that ability.  

It would be very slow if we did this by searching all the debug information for every inlined subroutine record looking for a call-site match, so I added one restriction to the call-site support.  This change will find all call sites for functions that also supply at least one line to the regular line table.  That way we can use the fact that the line table search will move the location to that subsequent line (but only within the same function).   When we find an actually moved source line match, we can search in the function that contained that line table entry for the call-site, and set the breakpoint location back to that.  

When I started writing tests for this new ability, it quickly became obvious that our support for virtual inline stepping was pretty buggy.  We didn't print the right file & line number for the breakpoint, and we didn't set the position in the "virtual inlined stack" correctly when we hit the breakpoint.  We also didn't step through the inlined frames correctly.  There was code to try to detect the right inlined stack position, but it had been refactored a while back with the comment that it was super confusing and the refactor was supposed to make it clearer, but the refactor didn't work either.

That code was made much clearer by abstracting the job of "handling the stack readjustment" to the various StopInfo's.  Previously, there was a big (and buggy) switch over stop info's.  Moving the responsibility to the stop info made this code much easier to reason about.

We also had no tests for virtual inlined stepping (our inlined stepping test was actually written specifically to avoid the formation of a virtual inlined stack...  So I also added tests for that along with the tests for setting the call-site breakpoints.

### CODE REVIEW

File: lldb/include/lldb/Breakpoint/BreakpointLocation.h [hunk 1]
```diff
@@ -281,6 +283,18 @@ class BreakpointLocation
 
   /// Returns the breakpoint location ID.
   lldb::break_id_t GetID() const { return m_loc_id; }
+  
+  // Set the line entry that should be shown to users for this location.
+  // It is up to the caller to verify that this is a valid entry to show.
+  // The current use of this is to distinguish among line entries from a
+  // virtual inlined call stack that all share the same address.
```
Comment (adrian-prantl) [inline lldb/include/lldb/Breakpoint/BreakpointLocation.h:290]: Nit: these should all be Doxygen comments using ///

File: lldb/include/lldb/Breakpoint/BreakpointLocation.h [hunk 2]
```diff
@@ -305,6 +319,16 @@ class BreakpointLocation
...
   bool IgnoreCountShouldStop();
+  
+  // If this location knows that the virtual stack frame it represents is
+  // not frame 0, return the suggested stack frame instead.  This will happen
+  // when the location's address contains a "virtual inlined call stack" and the
+  // breakpoint was set on a file & line that are not at the bottom of that
+  // stack.  For now we key off the "preferred line entry" - looking for that
+  // in the blocks that start with the stop PC.
+  // This version of the API doesn't take an "inlined" parameter because it
+  // only changes frames in the inline stack.
```
Comment (adrian-prantl) [inline lldb/include/lldb/Breakpoint/BreakpointLocation.h:330]: same here ...

File: lldb/include/lldb/Breakpoint/BreakpointLocation.h [hunk 3]
```diff
@@ -369,6 +393,13 @@ class BreakpointLocation
   lldb::break_id_t m_loc_id; ///< Breakpoint location ID.
   StoppointHitCounter m_hit_counter; ///< Number of times this breakpoint
                                      /// location has been hit.
+  std::optional<LineEntry> m_preferred_line_entry; // If this exists, use it to print the stop
```
Comment (adrian-prantl) [inline lldb/include/lldb/Breakpoint/BreakpointLocation.h]: IMO the end-of line comment style only makes sense if the comment fits on one line.
```suggestion
  /// If this exists, use it to print the stop description rather than the LineEntry
  /// ...
  std::optional<LineEntry> m_preferred_line_entry; 
```

Comment (adrian-prantl) [reply lldb/include/lldb/Breakpoint/BreakpointLocation.h]: And if you use the Doxygen comment syntax, IDEs will know to associate the comment with the following declaration.

File: lldb/source/Breakpoint/BreakpointLocation.cpp [hunk 4]
```diff
@@ -656,6 +671,49 @@ void BreakpointLocation::SendBreakpointLocationChangedEvent(
...
+
+    const InlineFunctionInfo *info = inlined_block->GetInlinedFunctionInfo();
+    if (info) {
+      if (preferred_decl == info->GetDeclaration())
+        return depth;
+      if (preferred_decl == info->GetCallSite())
+        return depth + 1;
+    }
+    inlined_block = inlined_block->GetInlinedParent();
+    depth++;
```
Comment (adrian-prantl) [inline lldb/source/Breakpoint/BreakpointLocation.cpp:712]: Should there be a `(if depth > 1000) return {}` condition to deal with completely broken debug info?

File: lldb/source/Symbol/CompileUnit.cpp [hunk 5]
```diff
@@ -312,6 +315,112 @@ void CompileUnit::ResolveSymbolContext(
...
+         &examine_block](Block &block) -> void {
+      // Iterate over the sibling child blocks of the incoming block.
+      Block *sibling_block = block.GetFirstChild();
+      while (sibling_block) {
+        // We only have to descend through the regular blocks, looking for
+        // immediate inlines, since those are the only ones that will have this
+        // callsite.
+        const InlineFunctionInfo *inline_info =
+            sibling_block->GetInlinedFunctionInfo();
+        if (inline_info) {
```
Comment (adrian-prantl) [inline lldb/source/Symbol/CompileUnit.cpp:356]:
```suggestion
        if (!inline_info)
          continue;
```

File: lldb/test/API/functionalities/inline-stepping/TestInlineStepping.py [hunk 6]
```diff
@@ -32,6 +32,12 @@ def test_step_in_template_with_python_api(self):
         self.build()
         self.step_in_template()
 
+    @add_test_categories(["pyapi"])
+    def test_virtual_inline_stepping(self):
+        """Test stepping through a virtual inlined call stack"""
+        self.build()
+        self.virtual_inline_stepping()
+
     def setUp(self):
```
Comment (adrian-prantl) [inline lldb/test/API/functionalities/inline-stepping/TestInlineStepping.py:41]: I believe this entire function is no longer needed.

File: lldb/test/API/functionalities/inline-stepping/TestInlineStepping.py [hunk 7]
```diff
@@ -357,3 +363,61 @@ def step_in_template(self):
...
+
+        # Make sure that the location is at the call site (run_to_line_breakpoint already asserted
+        # that there's one location.):
+        bkpt_loc = bkpt.location[0]
+        strm = lldb.SBStream()
+        result = bkpt_loc.GetDescription(strm, lldb.eDescriptionLevelFull)
+
+        self.assertTrue(result, "Got a location description")
+        desc = strm.GetData()
+        print(f"Description:\n{desc}\n")
```
Comment (adrian-prantl) [inline lldb/test/API/functionalities/inline-stepping/TestInlineStepping.py:384]: should this be `if self.TraceOn():`?

File: lldb/source/Breakpoint/BreakpointLocation.cpp (hunk 4 above)
Comment (jimingham) [reply lldb/source/Breakpoint/BreakpointLocation.cpp:712]: That would mean there were cycles in the blocks?  I don't think we guard against that anywhere else that we're traversing the block structure.

File: lldb/source/Symbol/CompileUnit.cpp [hunk 8]
```diff
@@ -312,6 +315,112 @@ void CompileUnit::ResolveSymbolContext(
...
+          // If this is the call-site we are looking for, record that:
+          // We need to be careful because the call site from the debug info
+          // will generally have a column, but the user might not have specified
+          // it.
+          Declaration found_decl = inline_info->GetCallSite();
+          uint32_t sought_column = sought_decl.GetColumn();
+          if (found_decl.FileAndLineEqual(sought_decl) &&
```
Comment (adrian-prantl) [inline lldb/source/Symbol/CompileUnit.cpp:363]:
```suggestion
          if (!found_decl.FileAndLineEqual(sought_decl))
            continue;
          if (sought_column != LLDB_INVALID_COLUMN_NUMBER &&
               sought_column != found_decl.GetColumn())
             continue;
```

File: lldb/source/Symbol/CompileUnit.cpp [hunk 9]
```diff
@@ -312,6 +315,112 @@ void CompileUnit::ResolveSymbolContext(
...
+              (sought_column == LLDB_INVALID_COLUMN_NUMBER ||
+               sought_column == found_decl.GetColumn())) {
+            // If we found a call site, it belongs not in this inlined block,
+            // but in the parent block that inlined it.
+            Address parent_start_addr;
+            if (sibling_block->GetParent()->GetStartAddress(
```
Comment (adrian-prantl) [inline lldb/source/Symbol/CompileUnit.cpp:369]: etc ...

File: lldb/test/API/functionalities/inline-stepping/TestInlineStepping.py (hunk 6 above)
Comment (jimingham) [reply lldb/test/API/functionalities/inline-stepping/TestInlineStepping.py:41]: Not sure what you mean.  This is the style we often write tests in where we have the to unittest significant `test_whatever` function be just a shim on the real test function.  That makes it easy to see what all the tests are up front without having to wade through the details.  That's how all the other tests in this file work as well.

**Review by adrian-prantl — APPROVED** [review]
Very exciting, this is something I've wished for for a long time!

File: lldb/test/API/functionalities/inline-stepping/TestInlineStepping.py (hunk 6 above)
Comment (adrian-prantl) [reply lldb/test/API/functionalities/inline-stepping/TestInlineStepping.py:41]: I was referring to the setUp() function that just calls the parent setUp function. If it's still needed ignore this,.

File: lldb/test/API/functionalities/inline-stepping/TestInlineStepping.py (hunk 7 above)
Comment (jimingham) [reply lldb/test/API/functionalities/inline-stepping/TestInlineStepping.py:384]: Debugging detritus, I'll just remove it.

File: lldb/source/Symbol/CompileUnit.cpp (hunk 5 above)
Comment (jimingham) [reply lldb/source/Symbol/CompileUnit.cpp:356]: I don't think that's right.  We still need to do the bit after the if block to recurse the children and set the next sibling block.  I'd have to have a "goto", you can't just continue here.

File: lldb/source/Breakpoint/BreakpointLocation.cpp [hunk 10]
```diff
@@ -508,8 +508,20 @@ void BreakpointLocation::GetDescription(Stream *s,
         s->PutCString("re-exported target = ");
       else
         s->PutCString("where = ");
+
+      // If there's a preferred line entry for printing, use that.
+      bool show_function_info = true;
+      if (GetPreferredLineEntry()) {
+        sc.line_entry = *GetPreferredLineEntry();
```
Comment (medismailben) [inline lldb/source/Breakpoint/BreakpointLocation.cpp:515]: nit: avoid 2 function calls
```suggestion
      if (auto may_be_line_entry = GetPreferredLineEntry()) {
        sc.line_entry = *may_be_line_entry;
```

File: lldb/source/Breakpoint/BreakpointLocation.cpp [hunk 11]
```diff
@@ -537,7 +549,10 @@ void BreakpointLocation::GetDescription(Stream *s,
         if (sc.line_entry.line > 0) {
           s->EOL();
           s->Indent("location = ");
-          sc.line_entry.DumpStopContext(s, true);
+          if (GetPreferredLineEntry())
+            GetPreferredLineEntry()->DumpStopContext(s, true);
```
Comment (medismailben) [inline lldb/source/Breakpoint/BreakpointLocation.cpp:553]: ditto

File: lldb/source/Breakpoint/BreakpointLocation.cpp [hunk 12]
```diff
@@ -656,6 +671,49 @@ void BreakpointLocation::SendBreakpointLocationChangedEvent(
   }
 }
 
+std::optional<uint32_t> BreakpointLocation::GetSuggestedStackFrameIndex() {
+  if (!GetPreferredLineEntry())
+    return {};
+  LineEntry preferred = *GetPreferredLineEntry();
```
Comment (medismailben) [inline lldb/source/Breakpoint/BreakpointLocation.cpp:677]: ditto

File: lldb/test/API/functionalities/inline-stepping/TestInlineStepping.py (hunk 6 above)
Comment (jimingham) [reply lldb/test/API/functionalities/inline-stepping/TestInlineStepping.py:41]: Ah.  The other tests still need the setUp.  I don't use need the work it does for the test I added, but it doesn't do very much, so IMO it's not worth making another test class just to avoid calling it.

File: lldb/source/Symbol/CompileUnit.cpp (hunk 9 above)
Comment (jimingham) [reply lldb/source/Symbol/CompileUnit.cpp:369]: Ditto.

File: lldb/source/Core/Declaration.cpp [hunk 13]
```diff
@@ -71,7 +71,7 @@ int Declaration::Compare(const Declaration &a, const Declaration &b) {
 }
 
 bool Declaration::FileAndLineEqual(const Declaration &declaration) const {
-  int file_compare = FileSpec::Compare(this->m_file, declaration.m_file, true);
+  int file_compare = FileSpec::Compare(this->m_file, declaration.m_file, false);
```
Comment (clayborg) [inline lldb/source/Core/Declaration.cpp:74]: Why was this changed?

File: lldb/source/Breakpoint/BreakpointLocation.cpp [hunk 14]
```diff
@@ -508,8 +508,20 @@ void BreakpointLocation::GetDescription(Stream *s,
...
         s->PutCString("where = ");
+
+      // If there's a preferred line entry for printing, use that.
+      bool show_function_info = true;
+      if (auto preferred = GetPreferredLineEntry()) {
+        sc.line_entry = *preferred;
+        // FIXME: We're going to get the function name wrong when the preferred
+        // line entry is not the lowest one.  For now, just leave the function
+        // out in this case, but we really should also figure out how to easily
+        // fake the function name here.
```
[the remark this answers is not in the document]
Comment (jimingham) [reply lldb/source/Breakpoint/BreakpointLocation.cpp:519]: What would it mean to replace the Module, the CompileUnit or the Function?  This is for a breakpoint location, so whatever we do has to result in this being the same address.  I was hesitant to introduce a more general "breakpoint gets to modify its symbol context" if I didn't know what it meant or how to ensure we didn't end up with a SymbolContext that wasn't at the breakpoint location's address.  I was originally planning to JUST store the "inlined call stack depth" to make it clear this couldn't change the PC value, but that ended up being more annoying that helpful.

File: lldb/source/Breakpoint/BreakpointLocation.cpp (hunk 12 above)
Comment (jimingham) [reply lldb/source/Breakpoint/BreakpointLocation.cpp]: Sure.

File: lldb/source/Breakpoint/BreakpointLocation.cpp (hunk 11 above)
Comment (jimingham) [reply lldb/source/Breakpoint/BreakpointLocation.cpp]: Sure

File: lldb/source/Symbol/CompileUnit.cpp (hunk 8 above)
Comment (jimingham) [reply lldb/source/Symbol/CompileUnit.cpp:363]: These can't continue either, they'd need a goto to get the sibling iteration & child recursion at the end of the while.

File: lldb/source/Breakpoint/BreakpointSite.cpp [hunk 15]
```diff
@@ -87,6 +87,22 @@ void BreakpointSite::GetDescription(Stream *s, lldb::DescriptionLevel level) {
   m_constituents.GetDescription(s, level);
 }
 
+std::optional<uint32_t> BreakpointSite::GetSuggestedStackFrameIndex() {
+
+  std::optional<uint32_t> result;
+  std::lock_guard<std::recursive_mutex> guard(m_constituents_mutex);
+  for (BreakpointLocationSP loc_sp : m_constituents.BreakpointLocations()) {
+    std::optional<uint32_t> this_result = loc_sp->GetSuggestedStackFrameIndex();
```
[the remark this answers is not in the document]
Comment (jimingham) [reply lldb/source/Breakpoint/BreakpointSite.cpp]: Sure.

File: lldb/source/Breakpoint/BreakpointSite.cpp [hunk 16]
```diff
@@ -87,6 +87,22 @@ void BreakpointSite::GetDescription(Stream *s, lldb::DescriptionLevel level) {
...
+    if (this_result) {
+      if (!result)
+        result = this_result;
+      else
+        result = std::max(*this_result, *result);
```
[the remark this answers is not in the document]
Comment (jimingham) [reply lldb/source/Breakpoint/BreakpointSite.cpp]: Sure.

File: lldb/source/Core/Declaration.cpp (hunk 13 above)
Comment (jimingham) [reply lldb/source/Core/Declaration.cpp]: Better to do it on the up-and-up.  I added a full parameter and set it in the two places this function is used.

File: lldb/source/Breakpoint/BreakpointLocation.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (jimingham) [reply lldb/source/Breakpoint/BreakpointLocation.cpp:519]: I added one more change here, since the intention is that you cannot change the PC of the breakpoint Location by setting a PreferredLineEntry, I changed the code to actually check that when you call SetPreferredLineEntry.  I put in an assert in the setter, so we can catch any violations of this in the testsuite, and then if this happens IRL, I made the BreakpointLocation log the event and ignore the setting.
